gh-128505: Expose an interface to sqlite3_file_control

hashbrowncipher opened the pull request:
sqlite3_file_control can affect the semantics of a sqlite3 database, but is not exposed by the sqlite.Connection class. In particular, this addition allows callers to use the [SQLITE_FCNTL_PERSIST_WAL](https://sqlite.org/c3ref/c_fcntl_begin_atomic_write.html#sqlitefcntlpersistwal) opcode. SQLITE_FCNTL_PERSIST_WAL ensures that callers can open a WAL-mode database on a read-only mount even when the writers have closed the DB.


<!-- gh-issue-number: gh-128505 -->
* Issue: gh-128505
<!-- /gh-issue-number -->

hashbrowncipher committed:
sqlite3_file_control

hashbrowncipher committed:
sqlite3_file_control

hashbrowncipher committed:
sqlite3_file_control

erlend-aasland commented on Lib/test/test_sqlite3/test_dbapi.py:
```diff
@@ -735,6 +736,30 @@ def test_database_keyword(self):
+
+
+    def test_file_control_raises(self):
+        with contextlib.closing(sqlite.connect(database=":memory:")) as cx:
```
You can use our `memory_database` helper here.

https://github.com/python/cpython/blob/ae23a012e6c8aadc4a588101cbe7bc86ede45627/Lib/test/test_sqlite3/util.py#L10-L12

erlend-aasland commented on Lib/test/test_sqlite3/test_dbapi.py:
```diff
@@ -1863,6 +1888,8 @@ def test_on_conflict_replace(self):
 
 
+
+
```
Please do not add unrelated changes, including style changes like this.

```suggestion
```

erlend-aasland reviewed (commented):
I briefly started looking at this; will do a full review later today (CET).

erlend-aasland commented on Misc/NEWS.d/next/Library/2025-01-05-00-55-41.gh-issue-128505.Nf5FY2.rst:
```diff
@@ -0,0 +1 @@
+sqlite Connection objects now expose a method set_file_control, which is a thin wrapper for `sqlite3_file_control https://www.sqlite.org/c3ref/file_control.html`_.
```
See https://devguide.python.org/documentation/markup/ for help with docs markup.

hashbrowncipher committed:
sqlite3_file_control

erlend-aasland commented on Modules/_sqlite/connection.c:
```diff
@@ -2173,6 +2173,45 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
 }
 
+/*[clinic input]
+_sqlite3.Connection.set_file_control as pysqlite_connection_set_file_control
```
`sqlite3_file_control` is an API for querying and setting values. `set_file_control` is misleading, since the API can also be used to query values. See similar API wrappers for `sqlite3_db_config` and `sqlite3_limit` in connection.c. How about making this a set/get pair: `setfilecontrol`/`getfilecontrol`? If we're adding it as a single API, let's go with `filecontrol`.

erlend-aasland commented on Modules/_sqlite/connection.c:
```diff
@@ -2173,6 +2173,45 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+    arg: long
+      argument to pass
+    /
+    dbname: str = NULL
```
Align this with the `name` keyword-only parameter of `_sqlite3.Connection.blobopen`.

```suggestion
    name: str = "main"
```

erlend-aasland commented on Modules/_sqlite/connection.c:
```diff
@@ -2173,6 +2173,45 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+_sqlite3.Connection.set_file_control as pysqlite_connection_set_file_control
+
+    op: int
+      a SQLITE_FCNTL_ constant
```
All parameter docstrings should be correctly written sentences; start with a capital letter and end with a punctuation mark.

erlend-aasland commented on Modules/_sqlite/connection.c:
```diff
@@ -2173,6 +2173,45 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+/*[clinic end generated code: output=d9d2d311892893b6 input=0253798d9514fea2]*/
+{
+    int rc;
+    long val = arg;
```
Not all `sqlite3_file_control` opcodes use the `int`[^1] datatype. For example `SQLITE_FCNTL_VFSNAME` and `SQLITE_FCNTL_TEMPFILENAME` are `char *`. Suggesting to limit this (like we do for `sqlite3_db_config`) to only the supported opcodes, and raise an exception for unsupported opcodes.

In any case, we should use `int` here.

```suggestion
    int val = arg;
```

[^1]: `int`, not `long`

erlend-aasland replied:
Please also address the first part of this remark.

hashbrowncipher replied:
oops! sorry for making you ask twice

hashbrowncipher replied:
Done. I have also removed the SQLITE_FCNTL_TEMPFILENAME and SQLITE_FCNTL_VFSNAME constants. I hope that's alright.

erlend-aasland commented on Modules/_sqlite/connection.c:
```diff
@@ -2173,6 +2173,45 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+    Py_END_ALLOW_THREADS
+
+    if (rc != SQLITE_OK) {
+        PyErr_SetString(self->ProgrammingError, sqlite3_errstr(rc));
```
`ProgrammingError` may not be the right fit for all opcodes. Ideally, we should use a variant of `_pysqlite_seterror` where we pass an error code instead of a SQLite database pointer. Perhaps I should make such an internal API available first.

hashbrowncipher replied:
I'd be happy to use any internal API that you build for this purpose.

erlend-aasland replied:
Give me a couple of days; my CPython bandwidth is pretty low this week, but I'll try to cook something up within the week. I'll try to jot down my plan in a dedicated issue soon.

erlend-aasland replied:
Draft PR up: #129929

hashbrowncipher committed:
sqlite3_file_control

hashbrowncipher committed:
sqlite3_file_control

erlend-aasland commented on Modules/_sqlite/connection.c:
```diff
@@ -2173,6 +2173,120 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
+
+    op: int
+      The SQLITE_FCNTL_* constant to invoke.
+    arg: long
```
Let's stick with `int`, since that's the datatype used by SQLite for these opcodes.

```suggestion
    arg: int
```

erlend-aasland commented on Misc/NEWS.d/next/Library/2025-01-05-00-55-41.gh-issue-128505.Nf5FY2.rst:
```diff
@@ -0,0 +1,3 @@
+sqlite Connection objects now expose a method
+:meth:`sqlite3.Connection.file_control`, which is a thin wrapper for
+`sqlite3_file_control <https://www.sqlite.org/c3ref/file_control.html>`_.
```
You also need to add docs for `file_control`. See `Doc/library/sqlite3.rst`.

hashbrowncipher committed:
change arg

hashbrowncipher committed:
remove unsupported opcodes

hashbrowncipher committed:
update docstring

hashbrowncipher committed:
add reference docs

erlend-aasland committed:
Pull in main

erlend-aasland committed:
gh-129928: Rework sqlite3 error helpers

Add a private API for raising DB-API compatible exceptions based on the
result code of SQLite C APIs. Some APIs do not store the error indicator
on the database pointer, so we need to be able to deduce the DB-API
compatible exception directly from the error code.

- rename _pysqlite_seterror() as set_error_from_db()
- introduce set_error_from_code()

hashbrowncipher committed:
use set_error_from_code

hashbrowncipher committed:
Merge remote-tracking branch 'erlend-aasland/sqlite/set-error' into sqlite3_file_control

hashbrowncipher committed:
fix test

erlend-aasland commented on Lib/test/test_sqlite3/test_dbapi.py:
```diff
@@ -724,6 +725,31 @@ def test_database_keyword(self):
+
+    def test_file_control_raises(self):
+        with memory_database() as cx:
+            with self.assertRaises(sqlite.InternalError):
```
`InternalError` sounds suspicious. What's the underlying SQLite error code? `SQLITE_MISUSE`?

hashbrowncipher replied:
Based on the failed test run, the full exception is `sqlite3.InternalError: unknown operation`, which appears to correspond to `SQLITE_NOTFOUND`.
